### PR TITLE
[WIP] What would a typed reconciler look like?

### PR DESCRIPTION
**Pull request by n3wscott**
Coding an example of a typed controller and an generated core reconciler other reconcilers can build off of.

### CODE REVIEW

**Comment by vaikas** [comment]
Is there some files omitted here? It's not clear to me where you link types to kinds that you want to reconcile. Also, seems like status update no longer happens?

**Comment by n3wscott** [comment]
There is some assumptions not documented yet. All changes to the status and finalizers of the passed in object to `ReconcileKind` will be propagated up to the object in the correct way.

**Comment by n3wscott** [comment]
Look in the "generated" files too...

File: pkg/reconciler/addressableservice/addressableservice.go [hunk 1]
```diff
@@ -18,115 +18,48 @@ package addressableservice
...
-	}
-	return reconcileErr
-}
+// Check that our Reconciler implements reconciler.Interface
+var _ Interface = (*Reconciler)(nil)
 
-func (r *Reconciler) reconcile(ctx context.Context, asvc *v1alpha1.AddressableService) error {
+// ReconcileKind implements addressableservice Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, asvc *v1alpha1.AddressableService) error {
 	if asvc.GetDeletionTimestamp() != nil {
```
Comment (mattmoor) [inline pkg/reconciler/addressableservice/addressableservice.go]: This is more boilerplate that we should elide.

Perhaps in addition to code-gen'ing `Interface` we can have a `Finalizer` interface with `FinalizeKind(ctx, asvc)`.

The signature to NewImpl would remain unchanged, but when the argument to `NewImpl`'s second argument implements this, it gets called?

File: pkg/reconciler/addressableservice/controller.go [hunk 2]
```diff
@@ -43,29 +32,21 @@ func NewController(
...
 		ServiceLister: svcInformer.Lister(),
-		Recorder: record.NewBroadcaster().NewRecorder(
-			scheme.Scheme, corev1.EventSource{Component: controllerAgentName}),
 	}
-	impl := controller.NewImpl(c, logger, "AddressableServices")
+	impl := NewImpl(ctx, r)
 
 	logger.Info("Setting up event handlers")
 
-	asInformer.Informer().AddEventHandler(controller.HandleAll(impl.Enqueue))
```
Comment (mattmoor) [inline pkg/reconciler/addressableservice/controller.go]: Setting this up automatically is a bad assumption, we used class-filtered enqueuing for Ingress and PodAutoscaler :(

**Comment by n3wscott** [comment]
/close
in favor of #167

**Pull request closed without merging**